langchain/src/prompts/pipeline.ts

```diff
@@ -0,0 +1,120 @@
+import { InputValues, PartialValues } from "../schema/index.js";
+import { BasePromptTemplate, BasePromptTemplateInput } from "./base.js";
+import { ChatPromptTemplate } from "./chat.js";
+import { SerializedBasePromptTemplate } from "./serde.js";
+
+export type PipelinePromptParams<
+  PromptTemplateType extends BasePromptTemplate
+> = {
+  name: string;
+  prompt: PromptTemplateType;
+};
+
+export type PipelinePromptTemplateInput<
+  PromptTemplateType extends BasePromptTemplate
+> = Omit<BasePromptTemplateInput, "inputVariables"> & {
+  pipelinePrompts: PipelinePromptParams<PromptTemplateType>[];
+  finalPrompt: PromptTemplateType;
+};
+
+export class PipelinePromptTemplate<
+  PromptTemplateType extends BasePromptTemplate
+> extends BasePromptTemplate {
+  pipelinePrompts: PipelinePromptParams<PromptTemplateType>[];
+
+  finalPrompt: PromptTemplateType;
+
+  constructor(input: PipelinePromptTemplateInput<PromptTemplateType>) {
+    super({ ...input, inputVariables: [] });
+    this.pipelinePrompts = input.pipelinePrompts;
+    this.finalPrompt = input.finalPrompt;
+    this.inputVariables = this.computeInputValues();
+  }
+
+  protected computeInputValues() {
+    const intermediateValues = this.pipelinePrompts.map(
+      (pipelinePrompt) => pipelinePrompt.name
+    );
+    const inputValues = this.pipelinePrompts
+      .map((pipelinePrompt) =>
+        pipelinePrompt.prompt.inputVariables.filter(
+          (inputValue) => !intermediateValues.includes(inputValue)
+        )
+      )
+      .flat();
+    return [...new Set(inputValues)];
+  }
+
+  protected static extractRequiredInputValues(
```

Comment:
Why protected on a static field?

Reply:
Just as documentation - that it's not intended to be part of the public interface